crates/nu-command/src/network/port.rs

```diff
@@ -121,25 +120,200 @@ fn get_free_port(
+            // - `size_pointer` still points to `size` from the first call.
+            let ret_code = unsafe { GetTcpTable2(Some(ptr), size_pointer, false) };
+            let ret_code = WIN32_ERROR(ret_code);
+            if ret_code != NO_ERROR {
+                // SAFETY:
+                // - `ptr` was allocated with `alloc(layout)` in this function.
+                // - Using the same `layout` to deallocate is correct.
+                unsafe { dealloc(ptr as *mut u8, layout) };
+                return Err(ret_code);
+            }
+
+            // SAFETY: `GetTcpTable2` returned `NO_ERROR`, so the header at `ptr` is initialized.
+            let header = unsafe { &*ptr };
+
+            // SAFETY:
+            // - Memory at `ptr` came from the global allocator and is initialized.
+            // - `TcpTable` is #[repr(C)] and layout-compatible with `MIB_TCPTABLE2` plus trailing rows.
+            // - We build a slice fat pointer only to carry the length; we do not dereference the slice itself here.
+            // - Casts between slice DSTs preserve the length metadata:
+            //     https://github.com/rust-lang/unsafe-code-guidelines/issues/288
+            //     https://github.com/rust-lang/reference/pull/1417
+            // - Casting to `*mut TcpTable` preserves that metadata for our DST.
+            // - `Box::from_raw` takes ownership and will free via the same allocator.
+            let table = unsafe {
+                let ptr = ptr::slice_from_raw_parts_mut(ptr, header.dwNumEntries as usize);
+                Box::from_raw(ptr as *mut TcpTable)
+            };
+
+            Ok(table)
+        }
+    }
+
+    /// Find an open port by checking the TCP table.
+    ///
+    /// On Windows, it is possible to bind to the same port multiple times if it was not
+    /// originally bound as an exclusive port[^so].
+    /// The Rust implementation of [`TcpListener::bind`] currently does not enforce exclusive
+    /// binding, which means the same port can be bound more than once.  
+    /// Because of this, we cannot simply try binding to a port to check if it is free.  
+    /// Instead, we query the [TCP table](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-gettcptable2)
+    /// to see which ports are already in use and then pick one that is not listed.
+    ///
+    /// [^so]: <https://docs.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse>
+    #[cfg(windows)]
+    pub fn search_port_in_range(
+        range: Spanned<RangeInclusive<u16>>,
+        call_span: Span,
+    ) -> Result<u16, ShellError> {
+        use std::collections::HashSet;
+
+        let table = TcpTable::new()
+            .map_err(|err| {
+                (
+                    err,
+                    match err {
+                        NO_ERROR => unreachable!("handled as Ok variant"),
+                        ERROR_INSUFFICIENT_BUFFER => "The buffer for TcpTable is not large enough",
```

Comment:
Not sure how likely it is to run into this with a TOCTOU when you are running it while a bunch of new connections are established.

Reply:
I'm not sure either but if you're worried, you as a user can still run this command in try loop I guess.